src/main/java/org/springframework/data/repository/config/SelectionSet.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.repository.config;
+
+import java.util.Collection;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Allows filtering of a collection in order to select a unique element. Once a unique element is found all further
+ * filters are ignored.
+ *
+ * @author Jens Schauder
+ */
+class SelectionSet<T> {
```

Comment:
I am wondering whether we shouldn't let `uniqueResult()` return an `Optional<T>` so that the client is forced to deal with the absence of the a value. Currently in that case we just throw an `IllegalStateException` saying "More then one element in collection." which doesn't give a lot of context. I'd rather let the client — who has that context — be forced to mitigate between the `Optional<T>` and `T` as it know what exactly the absence means and can craft a way more meaningful exception. With the current state, the client basically assumes the call would produce something reasonable and moves on.